{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Remember to reverse loop for negative strand. From iteration to beginning\n",
    "### Are inclusion and exclusion NMD algs different?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import nmd\n",
    "import gffutils\n",
    "import itertools"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "gffutils.create_db??"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "database='/Users/rhythmicstar/projects/exon_evolution//gencode.v19.annotation.outrigger.gtf.db'\n",
    "nmd_file='/Users/rhythmicstar/ipython_notebook/nmd/example_splicing.bed'\n",
    "event_ids_df = nmd.possible_nmd(nmd_file)\n",
    "DBHuman = gffutils.FeatureDB(database)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>isoform1</th>\n",
       "      <th>isoform2</th>\n",
       "      <th>exon</th>\n",
       "      <th>junction</th>\n",
       "      <th>strand</th>\n",
       "      <th>type</th>\n",
       "      <th>inclusion_nmd</th>\n",
       "      <th>exclusion_nmd</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr16:29383506-29484368</td>\n",
       "      <td>chr16:29383506-29383883</td>\n",
       "      <td>chr16:29383884-29383940</td>\n",
       "      <td>chr16:29383941-29484368</td>\n",
       "      <td>+</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr9:137102499-137103363</td>\n",
       "      <td>chr9:137102499-137103132</td>\n",
       "      <td>chr9:137103133-137103170</td>\n",
       "      <td>chr9:137103171-137103363</td>\n",
       "      <td>+</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr16:21404089-21404856</td>\n",
       "      <td>chr16:21404089-21404157</td>\n",
       "      <td>chr16:21404158-21404184</td>\n",
       "      <td>chr16:21404185-21404856</td>\n",
       "      <td>+</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr2:97198492-97209680</td>\n",
       "      <td>chr2:97198492-97200333</td>\n",
       "      <td>chr2:97200334-97200362</td>\n",
       "      <td>chr2:97200363-97209680</td>\n",
       "      <td>+</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr10:12149962-12156906</td>\n",
       "      <td>chr10:12149962-12155777</td>\n",
       "      <td>chr10:12155778-12155931</td>\n",
       "      <td>chr10:12155932-12156906</td>\n",
       "      <td>+</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr6:138921125-138929976</td>\n",
       "      <td>chr6:138926482-138929976</td>\n",
       "      <td>chr6:138926401-138926481</td>\n",
       "      <td>chr6:138921125-138926400</td>\n",
       "      <td>-</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr9:112694234-112716460</td>\n",
       "      <td>chr9:112701881-112716460</td>\n",
       "      <td>chr9:112701788-112701880</td>\n",
       "      <td>chr9:112694234-112701787</td>\n",
       "      <td>-</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chr20:63700703-63702135</td>\n",
       "      <td>chr20:63701896-63702135</td>\n",
       "      <td>chr20:63701830-63701895</td>\n",
       "      <td>chr20:63700703-63701829</td>\n",
       "      <td>-</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>chr9:137102898-137103321</td>\n",
       "      <td>chr9:137102898-137103019</td>\n",
       "      <td>chr9:137103020-137103057</td>\n",
       "      <td>chr9:137103058-137103321</td>\n",
       "      <td>+</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>chr14:50161610-50180571</td>\n",
       "      <td>chr14:50174553-50180571</td>\n",
       "      <td>chr14:50174454-50174552</td>\n",
       "      <td>chr14:50161610-50174453</td>\n",
       "      <td>-</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>chr12:10761697-10761982</td>\n",
       "      <td></td>\n",
       "      <td>+</td>\n",
       "      <td>inclusion</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>chr3:186502751-186502890</td>\n",
       "      <td></td>\n",
       "      <td>+</td>\n",
       "      <td>inclusion</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>chr3:186506099-186506205</td>\n",
       "      <td></td>\n",
       "      <td>+</td>\n",
       "      <td>inclusion</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>chr12:10761697-10761982</td>\n",
       "      <td></td>\n",
       "      <td>-</td>\n",
       "      <td>inclusion</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>chr3:186502353-186502486</td>\n",
       "      <td></td>\n",
       "      <td>+</td>\n",
       "      <td>inclusion</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    isoform1                  isoform2  \\\n",
       "0    chr16:29383506-29484368   chr16:29383506-29383883   \n",
       "1   chr9:137102499-137103363  chr9:137102499-137103132   \n",
       "2    chr16:21404089-21404856   chr16:21404089-21404157   \n",
       "3     chr2:97198492-97209680    chr2:97198492-97200333   \n",
       "4    chr10:12149962-12156906   chr10:12149962-12155777   \n",
       "5   chr6:138921125-138929976  chr6:138926482-138929976   \n",
       "6   chr9:112694234-112716460  chr9:112701881-112716460   \n",
       "7    chr20:63700703-63702135   chr20:63701896-63702135   \n",
       "8   chr9:137102898-137103321  chr9:137102898-137103019   \n",
       "9    chr14:50161610-50180571   chr14:50174553-50180571   \n",
       "10                                                       \n",
       "11                                                       \n",
       "12                                                       \n",
       "13                                                       \n",
       "14                                                       \n",
       "\n",
       "                        exon                  junction strand       type  \\\n",
       "0    chr16:29383884-29383940   chr16:29383941-29484368      +              \n",
       "1   chr9:137103133-137103170  chr9:137103171-137103363      +              \n",
       "2    chr16:21404158-21404184   chr16:21404185-21404856      +              \n",
       "3     chr2:97200334-97200362    chr2:97200363-97209680      +              \n",
       "4    chr10:12155778-12155931   chr10:12155932-12156906      +              \n",
       "5   chr6:138926401-138926481  chr6:138921125-138926400      -              \n",
       "6   chr9:112701788-112701880  chr9:112694234-112701787      -              \n",
       "7    chr20:63701830-63701895   chr20:63700703-63701829      -              \n",
       "8   chr9:137103020-137103057  chr9:137103058-137103321      +              \n",
       "9    chr14:50174454-50174552   chr14:50161610-50174453      -              \n",
       "10   chr12:10761697-10761982                                +  inclusion   \n",
       "11  chr3:186502751-186502890                                +  inclusion   \n",
       "12  chr3:186506099-186506205                                +  inclusion   \n",
       "13   chr12:10761697-10761982                                -  inclusion   \n",
       "14  chr3:186502353-186502486                                +  inclusion   \n",
       "\n",
       "   inclusion_nmd exclusion_nmd  \n",
       "0                               \n",
       "1                               \n",
       "2                               \n",
       "3                               \n",
       "4                               \n",
       "5                               \n",
       "6                               \n",
       "7                               \n",
       "8                               \n",
       "9                               \n",
       "10                              \n",
       "11                              \n",
       "12                              \n",
       "13                              \n",
       "14                              "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "event_ids_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def trans_exon_set(exon):\n",
    "    transcripts_with_exon = set()\n",
    "    for exon_transc in DBHuman.parents(exon, featuretype=\"transcript\"):\n",
    "        transcripts_with_exon.add(exon_transc['transcript_id'][0])\n",
    "    return transcripts_with_exon #parent_transcripts_of_exon\n",
    "\n",
    "def all_trans_set(exon):\n",
    "    all_transcripts = set()\n",
    "    for transc in DBHuman.children(exon['gene_id'][0], featuretype=\"transcript\"):\n",
    "        all_transcripts.add(transc['transcript_id'][0])\n",
    "    return all_transcripts #call it transcripts_from_gene_containing_exon\n",
    "\n",
    "def create_dict(all_transcripts, strand):\n",
    "    transcript_exons = dict()\n",
    "    for transcript_id in all_transcripts:\n",
    "        this_transcript_exons = tuple(DBHuman.children(transcript_id, featuretype = \"exon\", order_by=\"start\"))\n",
    "        #set reverse (look up) = true\n",
    "        internal_exons = this_transcript_exons[1:len(this_transcript_exons)-1]\n",
    "        transcript_cds = tuple(DBHuman.children(transcript_id, featuretype='CDS', order_by=\"start\"))\n",
    "        for startcodon in DBHuman.children(transcript_id, featuretype='start_codon'):\n",
    "            if (strand=='+'):\n",
    "                firstcds = transcript_cds[0:1]\n",
    "                real_transcript_cds = ((startcodon,) + (firstcds) + internal_exons)\n",
    "            else:\n",
    "                firstcds = transcript_cds[-1]\n",
    "                real_transcript_cds = (internal_exons + (firstcds,) + (startcodon,)) #can use reverse order of internal exons\n",
    "            #print(\"real_transcript_cds\", real_transcript_cds)\n",
    "            transcript_exons[transcript_id] = real_transcript_cds\n",
    "    return transcript_exons\n",
    "\n",
    "    \n",
    "def inclusion_possible_nmd(trans_without_exon, trans_with_exon, strand):\n",
    "    trans_with_exon_set = set(trans_with_exon)\n",
    "    trans_without_exon_set = set(trans_without_exon)\n",
    "    as_exon = tuple(trans_with_exon_set-trans_without_exon_set)[0]\n",
    "    #print(as_exon)\n",
    "    index = trans_with_exon.index(as_exon)\n",
    "    if (strand=='+'):\n",
    "        for iterator in range(index, len(trans_with_exon)):\n",
    "            for stop in DBHuman.children(trans_with_exon[iterator]['gene_id'][0], featuretype='stop_codon'):\n",
    "                #print(\"stop codon: \", stop)\n",
    "                if stop_codon_nmd(trans_with_exon[index], stop):\n",
    "                    return True\n",
    "    if (strand=='-'):\n",
    "        for iterator in range(0, index+1):\n",
    "            for stop in DBHuman.children(trans_with_exon[iterator]['gene_id'][0], featuretype='stop_codon'):\n",
    "                #print(\"stop codon: \", stop)\n",
    "                if stop_codon_nmd(trans_with_exon[index], stop):\n",
    "                    return True\n",
    "    return False\n",
    "\n",
    "\n",
    "def exclusion_possible_nmd(trans_without_exon, trans_with_exon, strand):\n",
    "    with open(\"without_exon.gtf\", \"w\") as f:\n",
    "        for line in trans_without_exon:\n",
    "            f.write(str(line))\n",
    "    with open(\"with_exon.gtf\", \"w\") as f:\n",
    "        for line in trans_with_exon:\n",
    "            f.write(str(line))\n",
    "    print(\"trans_with_exon: \", trans_with_exon)\n",
    "    print(\"trans_without_exon: \", trans_without_exon)\n",
    "    trans_with_exon_set = set(trans_with_exon)\n",
    "    trans_without_exon_set = set(trans_without_exon)\n",
    "    as_exon = tuple(trans_with_exon_set-trans_without_exon_set)[0]\n",
    "    print(\"as_exon: \", as_exon)\n",
    "    #print(as_exon)\n",
    "    index = trans_with_exon.index(as_exon)\n",
    "    if (strand=='+'):\n",
    "#       print(\"trans_without_exon: \", trans_without_exon)\n",
    "        for iterator in range(index, len(trans_without_exon)+1):\n",
    "            for stop in DBHuman.children(trans_without_exon[iterator]['gene_id'][0], featuretype='stop_codon'):\n",
    "#               print(\"stop codon: \", stop)\n",
    "                if stop_codon_nmd(trans_without_exon[index], stop):\n",
    "                    return True\n",
    "    else:\n",
    "        for iterator in range(0, index):\n",
    "            for stop in DBHuman.children(trans_with_exon[iterator]['gene_id'][0], featuretype='stop_codon'):\n",
    "                #print(\"stop codon: \", stop)\n",
    "                if stop_codon_nmd(trans_with_exon[index], stop):\n",
    "                    return True\n",
    "    return False\n",
    "    \n",
    "def stop_codon_nmd(human_exon, stop_codon):\n",
    "    if stop_codon.strand == '+':\n",
    "        if stop_codon.start >= human_exon.start and stop_codon.end <= \\\n",
    "                human_exon.end and stop_codon.end + 50 < human_exon.end:\n",
    "            return True\n",
    "    else:\n",
    "        if stop_codon.start >= human_exon.start and stop_codon.end <= \\\n",
    "                human_exon.end and human_exon.start + 50 < stop_codon.start:\n",
    "            return True\n",
    "    return False"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {
    "collapsed": false,
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "-----------------\n",
      "chr16:29383884-29383940\n",
      "exon:chr16:29383884-29383940:+\n",
      "-----------------\n",
      "chr9:137103133-137103170\n",
      "exon:chr9:137103133-137103170:+\n",
      "-----------------\n",
      "chr16:21404158-21404184\n",
      "exon:chr16:21404158-21404184:+\n",
      "-----------------\n",
      "chr2:97200334-97200362\n",
      "exon:chr2:97200334-97200362:+\n",
      "-----------------\n",
      "chr10:12155778-12155931\n",
      "exon:chr10:12155778-12155931:+\n",
      "-----------------\n",
      "chr6:138926401-138926481\n",
      "exon:chr6:138926401-138926481:-\n",
      "-----------------\n",
      "chr9:112701788-112701880\n",
      "exon:chr9:112701788-112701880:-\n",
      "-----------------\n",
      "chr20:63701830-63701895\n",
      "exon:chr20:63701830-63701895:-\n",
      "-----------------\n",
      "chr9:137103020-137103057\n",
      "exon:chr9:137103020-137103057:+\n",
      "-----------------\n",
      "chr14:50174454-50174552\n",
      "exon:chr14:50174454-50174552:-\n",
      "-----------------\n",
      "chr12:10761697-10761982\n",
      "exon:chr12:10761697-10761982:+\n",
      "-----------------\n",
      "chr3:186502751-186502890\n",
      "exon:chr3:186502751-186502890:+\n",
      "trans_with_exon:  (<Feature start_codon (chr3:186501400-186501402[+]) at 0x1066a5898>, <Feature CDS (chr3:186501400-186501428[+]) at 0x10660e748>, <Feature exon (chr3:186502218-186502266[+]) at 0x106626e48>, <Feature exon (chr3:186502353-186502485[+]) at 0x106626668>, <Feature exon (chr3:186502751-186502890[+]) at 0x1066a15c0>, <Feature exon (chr3:186503672-186503840[+]) at 0x1066a1470>, <Feature exon (chr3:186503953-186504062[+]) at 0x1066a1a20>, <Feature exon (chr3:186504291-186504434[+]) at 0x1067b00b8>, <Feature exon (chr3:186504916-186505053[+]) at 0x1066b6dd8>, <Feature exon (chr3:186505284-186505373[+]) at 0x1066b6438>, <Feature exon (chr3:186505592-186505671[+]) at 0x1066b6160>)\n",
      "trans_without_exon:  (<Feature start_codon (chr3:186501400-186501402[+]) at 0x1067672b0>, <Feature CDS (chr3:186501400-186501428[+]) at 0x106689d30>, <Feature exon (chr3:186502218-186502266[+]) at 0x106626f98>, <Feature exon (chr3:186502353-186502485[+]) at 0x106626470>, <Feature exon (chr3:186503672-186503840[+]) at 0x106626b00>, <Feature exon (chr3:186503953-186504062[+]) at 0x106626fd0>, <Feature exon (chr3:186504291-186504434[+]) at 0x1067676a0>, <Feature exon (chr3:186504916-186505053[+]) at 0x106767be0>, <Feature exon (chr3:186505284-186505373[+]) at 0x106767e48>, <Feature exon (chr3:186505592-186505671[+]) at 0x1066339e8>)\n",
      "as_exon:  chr3\tHAVANA\texon\t186502751\t186502890\t.\t+\t.\texon_id \"ENSE00002462331.1\"; transcript_name \"EIF4A2-003,EIF4A2-001,EIF4A2-018,EIF4A2-004\"; gene_type \"protein_coding\"; gene_name \"EIF4A2\"; gene_id \"ENSG00000156976.10\"; havana_gene \"OTTHUMG00000156564.2\"; gene_status \"KNOWN\"; transcript_status \"PUTATIVE,KNOWN\"; havana_transcript \"OTTHUMT00000344609.1,OTTHUMT00000344612.1,OTTHUMT00000344626.2,OTTHUMT00000344611.1\"; exon_number \"4,3\"; transcript_id \"ENST00000323963.5,ENST00000440191.2,ENST00000498746.1,ENST00000425053.1\"; level \"2\"; ccdsid \"CCDS3282.1\"; transcript_type \"protein_coding,nonsense_mediated_decay\"; location_id \"exon:chr3:186502751-186502890:+\"; tag \"appris_principal,mRNA_end_NF,basic,CCDS,cds_end_NF\";\n",
      "True\n",
      "inc set()\n",
      "exc {'ENST00000443963.1'}\n",
      "trans_with_exon:  (<Feature start_codon (chr3:186501400-186501402[+]) at 0x10663d9e8>, <Feature CDS (chr3:186501400-186501428[+]) at 0x106632a90>, <Feature exon (chr3:186502221-186502266[+]) at 0x1066a58d0>, <Feature exon (chr3:186502353-186502485[+]) at 0x1066a5a20>, <Feature exon (chr3:186502751-186502890[+]) at 0x106683da0>, <Feature exon (chr3:186503672-186503840[+]) at 0x10660ce10>, <Feature exon (chr3:186503953-186504062[+]) at 0x10660c9e8>, <Feature exon (chr3:186504291-186504434[+]) at 0x1066f77b8>, <Feature exon (chr3:186504916-186505053[+]) at 0x1066f7a58>, <Feature exon (chr3:186505284-186505373[+]) at 0x1067352e8>, <Feature exon (chr3:186505592-186505671[+]) at 0x1067354a8>)\n",
      "trans_without_exon:  (<Feature start_codon (chr3:186501400-186501402[+]) at 0x1066a5828>, <Feature CDS (chr3:186501400-186501428[+]) at 0x1066a5a90>, <Feature exon (chr3:186502221-186502266[+]) at 0x106626438>, <Feature exon (chr3:186502353-186502485[+]) at 0x1066263c8>, <Feature exon (chr3:186503672-186503840[+]) at 0x106629e80>, <Feature exon (chr3:186503953-186504062[+]) at 0x10660ac88>, <Feature exon (chr3:186504291-186504434[+]) at 0x1066067f0>, <Feature exon (chr3:186504916-186505053[+]) at 0x1066069b0>, <Feature exon (chr3:186505284-186505373[+]) at 0x106606ba8>, <Feature exon (chr3:186505592-186505671[+]) at 0x10660e4e0>)\n",
      "as_exon:  chr3\tHAVANA\texon\t186502751\t186502890\t.\t+\t.\texon_id \"ENSE00002462331.1\"; transcript_name \"EIF4A2-003,EIF4A2-001,EIF4A2-018,EIF4A2-004\"; gene_type \"protein_coding\"; gene_name \"EIF4A2\"; gene_id \"ENSG00000156976.10\"; havana_gene \"OTTHUMG00000156564.2\"; gene_status \"KNOWN\"; transcript_status \"PUTATIVE,KNOWN\"; havana_transcript \"OTTHUMT00000344609.1,OTTHUMT00000344612.1,OTTHUMT00000344626.2,OTTHUMT00000344611.1\"; exon_number \"4,3\"; transcript_id \"ENST00000323963.5,ENST00000440191.2,ENST00000498746.1,ENST00000425053.1\"; level \"2\"; ccdsid \"CCDS3282.1\"; transcript_type \"protein_coding,nonsense_mediated_decay\"; location_id \"exon:chr3:186502751-186502890:+\"; tag \"appris_principal,mRNA_end_NF,basic,CCDS,cds_end_NF\";\n",
      "True\n",
      "inc set()\n",
      "exc {'ENST00000443963.1', 'ENST00000426808.1'}\n",
      "-----------------\n",
      "chr3:186506099-186506205\n",
      "exon:chr3:186506099-186506205:+\n",
      "inc {'ENST00000425053.1'}\n",
      "exc set()\n",
      "-----------------\n",
      "chr12:10761697-10761982\n",
      "exon:chr12:10761697-10761982:-\n",
      "inc {'ENST00000544850.1'}\n",
      "exc set()\n",
      "inc {'ENST00000544850.1', 'ENST00000540074.1'}\n",
      "exc set()\n",
      "-----------------\n",
      "chr3:186502353-186502486\n",
      "exon:chr3:186502353-186502486:+\n"
     ]
    }
   ],
   "source": [
    "for i, row in event_ids_df.iterrows():\n",
    "    print('-----------------')\n",
    "    print(row['exon'])\n",
    "    inclusion_exons = set()\n",
    "    exclusion_exons = set()\n",
    "    location = 'exon:' + row['exon'] + \":\" + row['strand']\n",
    "    print(location)\n",
    "    try:\n",
    "        exon = DBHuman[location]\n",
    "        transcripts_with_exon = trans_exon_set(exon)\n",
    "        all_transcripts = all_trans_set(exon)\n",
    "        transcripts_without_exon = all_transcripts-transcripts_with_exon\n",
    "        transcript_exons = create_dict(all_transcripts, row['strand'])\n",
    "\n",
    "        for t1 in transcripts_without_exon:\n",
    "            try:\n",
    "                for t2 in transcripts_with_exon:\n",
    "                    t1_ex = transcript_exons[t1]\n",
    "                    t2_ex = transcript_exons[t2]\n",
    "#                     print(\"t1\", t1_ex)\n",
    "                    if ((row['strand']=='+' and t1_ex[1] == t2_ex[1]) or (row['strand']=='-' and t1_ex[-1] == t2_ex[-1])):\n",
    "                        if len(set(t2_ex).symmetric_difference(set(t1_ex)))==1:\n",
    "                            if (DBHuman[t1]['transcript_type'][0]!='nonsense_mediated_decay') & (DBHuman[t2]['transcript_type'][0]=='nonsense_mediated_decay'):\n",
    "                                if inclusion_possible_nmd(t1_ex, t2_ex, row['strand']):\n",
    "                                    inclusion_exons.add(t2)\n",
    "                            if (DBHuman[t1]['transcript_type'][0]=='nonsense_mediated_decay') & (DBHuman[t2]['transcript_type'][0]!='nonsense_mediated_decay'):\n",
    "                                if exclusion_possible_nmd(t1_ex, t2_ex, row['strand']):\n",
    "                                    print(\"True\")\n",
    "                                    exclusion_exons.add(t1)\n",
    "                            print(\"inc\", inclusion_exons)\n",
    "                            print(\"exc\", exclusion_exons)\n",
    "\n",
    "            except KeyError:\n",
    "                t1_ex=None\n",
    "    except gffutils.FeatureNotFoundError:\n",
    "        exon = None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
