{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#imports\n",
    "#for gffutils database\n",
    "import gffutils\n",
    "\n",
    "#for pandas data frame\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Make df with splicing data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "def possible_nmd(database, nmd_file):    \n",
    "    splicing_data = pd.read_csv(nmd_file, header=None, sep='\\s+')\n",
    "    index = pd.Index(splicing_data[3])\n",
    "    event_ids = pd.Series(index, name='event_id')\n",
    "    event_ids_df = event_ids.to_frame()\n",
    "    event_ids_df['isoform1'] = splicing_data[3].map(lambda x: x.split('|')[0].split('junction:')[1][:-2])\n",
    "    event_ids_df['isoform2'] = splicing_data[3].map(lambda x: x.split('isoform2=junction:')[1].split('@')[0][:-2])\n",
    "    event_ids_df['exon'] = splicing_data[3].map(lambda x: x.split('exon:')[1].split('@')[0][:-2])\n",
    "    event_ids_df['junction'] = splicing_data[3].map(lambda x: x.split('@junction:')[1].split(' ')[0][:-2])\n",
    "    event_ids_df['inclusion_nmd'] = ''\n",
    "    event_ids_df['exclusion_nmd'] = ''\n",
    "    event_ids_df = event_ids_df[['isoform1', 'isoform2', 'exon', 'junction', 'inclusion_nmd', 'exclusion_nmd']]\n",
    "\n",
    "    pos_str = pd.DataFrame([['', '', 'chr3:186506099-186506205', '', '', '']], columns=list(['isoform1','isoform2','exon', 'junction', 'inclusion_nmd', 'exclusion_nmd']))\n",
    "    neg_str = pd.DataFrame([['', '', 'chr8:109213972-109260959', '', '', '']], columns=list(['isoform1','isoform2','exon', 'junction', 'inclusion_nmd', 'exclusion_nmd']))\n",
    "    d = pd.DataFrame([['', '', 'chr2:191745547-191830270', '', '', '']], columns=list(['isoform1','isoform2','exon', 'junction', 'inclusion_nmd', 'exclusion_nmd']))\n",
    "    e = pd.DataFrame([['', '', 'chr3:196662273-196669464', '', '', '']], columns=list(['isoform1','isoform2','exon', 'junction', 'inclusion_nmd', 'exclusion_nmd']))\n",
    "    f = pd.DataFrame([['', '', 'chr9:100395705-100436029', '', '', '']], columns=list(['isoform1','isoform2','exon', 'junction', 'inclusion_nmd', 'exclusion_nmd']))\n",
    "    g = pd.DataFrame([['', '', 'chr8:101715144-101734315', '', '', '']], columns=list(['isoform1','isoform2','exon', 'junction', 'inclusion_nmd', 'exclusion_nmd']))\n",
    "    \n",
    "    asout_df = pd.DataFrame([['', '', 'chr3:186502751-186502891', '', '', '']], columns=list(['isoform1','isoform2','exon', 'junction', 'inclusion_nmd', 'exclusion_nmd']))    \n",
    "\n",
    "    event_ids_df = event_ids_df.append([pos_str, neg_str, d, e, f, g, asout_df], ignore_index=True)\n",
    "    return event_ids_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "collapsed": false,
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "def include_exon_nmd(event_ids_df):\n",
    "    DBHuman = gffutils.FeatureDB(database, keep_order=True)\n",
    "    for i, row in event_ids_df.iterrows():\n",
    "        nmd = False\n",
    "        location = row['exon']\n",
    "        for index, human_exon in enumerate(DBHuman.features_of_type(('exon'), limit=location)):\n",
    "            if (human_exon['transcript_type']==[\"nonsense_mediated_decay\"]):\n",
    "                for stop_codon in DBHuman.children(human_exon['gene_id'][0], featuretype='stop_codon'):\n",
    "                    if (stop_codon_nmd(human_exon, stop_codon)==True):\n",
    "                        nmd = True\n",
    "        if (nmd==True):\n",
    "            event_ids_df.ix[i, 4] = 'True'\n",
    "        else:\n",
    "            event_ids_df.ix[i, 4] = 'False'\n",
    "    return event_ids_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# def exclude_exon_nmd(event_ids_df):\n",
    "#     DBHuman = gffutils.FeatureDB(database, keep_order=True)\n",
    "#     for i, row in event_ids_df.iterrows():\n",
    "#         nmd = False\n",
    "#         location = row['exon']\n",
    "#         #print('exon:', location)\n",
    "#         for exon in (DBHuman.features_of_type(('exon'), limit=location)):\n",
    "#             if (location==\"{chrom}:{start}-{end}\".format(chrom=human_exon.chrom, start=human_exon.start, end=human_exon.end):\n",
    "#                 for transcript in DBHuman.children(human_exon['gene_id'][0], featuretype='transcript'): #find transcripts that exon is not in\n",
    "#                     for human_exon in DBHuman.children(human_exon['transcript'][0], featuretype='exon'): #look in all exons after the missing exon\n",
    "#                         if (human_exon['transcript_type']==[\"nonsense_mediated_decay\"]):    \n",
    "#                             for stop_codon in DBHuman.children(human_exon['gene_id'][0], featuretype='stop_codon'):\n",
    "#                                 if (exon.strand == '+' and (exon.start >= event_ids_df.ix[row, 'exon'].split(':')[1])): #if strand is +, all exons with bigger location\n",
    "#                                     if (stop_codon_nmd(human_exon, stop_codon)==True):\n",
    "#                                         nmd = True\n",
    "#                                 else if (exon.strand == '-' and (exon.end <= event_ids_df.ix[row, 'exon'].split(':')[1])): #if strand is -, all exons with smaller location\n",
    "#                                     if (stop_codon_nmd(human_exon, stop_codon)==True):\n",
    "#                                         nmd = True\n",
    "#                 break\n",
    "#         if (nmd==True):\n",
    "#             event_ids_df.ix[i, 5] = 'True'\n",
    "#             event_ids_df.ix[i, 5] = 'False'\n",
    "#     return event_ids_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def stop_codon_nmd(human_exon, stop_codon):\n",
    "    if stop_codon.strand == '+':\n",
    "        if (stop_codon.start>=human_exon.start and stop_codon.end<=human_exon.end and stop_codon.end+50<human_exon.end):\n",
    "#             print(\"\\t\\t\\t\\tExon ({exon}) Start: {start}, End: {end}, is an NMD exon\".format(exon=human_exon.id, start=human_exon.start, end=human_exon.end))\n",
    "#             human_exon_id = str(human_exon['exon_id']).split(\"'\")[1]\n",
    "            return True\n",
    "    if stop_codon.strand == '-':\n",
    "        if (stop_codon.start>=human_exon.start and stop_codon.end<=human_exon.end and human_exon.start+50<stop_codon.start):\n",
    "#             print(\"\\t\\t\\t\\tExon ({exon}) Start: {start}, End: {end}, is an NMD exon\".format(exon=human_exon.id, start=human_exon.start, end=human_exon.end))\n",
    "#             human_exon_id = str(human_exon['exon_id']).split(\"'\")[1]\n",
    "            return True\n",
    "    return False"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "def main(database, nmd_file):\n",
    "    database='/Users/rhythmicstar/projects/exon_evolution//gencode.v19.annotation.gtf.db'\n",
    "    nmd_file='/Users/rhythmicstar/ipython_notebook/nmd/example_splicing.bed'\n",
    "    event_ids_df = possible_nmd(database, nmd_file)\n",
    "    #if negative correlation\n",
    "    event_ids_df = include_exon_nmd(event_ids_df)\n",
    "    #if positive correlation\n",
    "    #event_ids_df = exclude_exon_nmd(event_ids_df)\n",
    "    return event_ids_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                    isoform1                  isoform2  \\\n",
      "0    chr16:29383506-29484368   chr16:29383506-29383883   \n",
      "1   chr9:137102499-137103363  chr9:137102499-137103132   \n",
      "2    chr16:21404089-21404856   chr16:21404089-21404157   \n",
      "3     chr2:97198492-97209680    chr2:97198492-97200333   \n",
      "4    chr10:12149962-12156906   chr10:12149962-12155777   \n",
      "5   chr6:138921125-138929976  chr6:138926482-138929976   \n",
      "6   chr9:112694234-112716460  chr9:112701881-112716460   \n",
      "7    chr20:63700703-63702135   chr20:63701896-63702135   \n",
      "8   chr9:137102898-137103321  chr9:137102898-137103019   \n",
      "9    chr14:50161610-50180571   chr14:50174553-50180571   \n",
      "10                                                       \n",
      "11                                                       \n",
      "12                                                       \n",
      "13                                                       \n",
      "14                                                       \n",
      "15                                                       \n",
      "16                                                       \n",
      "\n",
      "                        exon                  junction inclusion_nmd  \\\n",
      "0    chr16:29383884-29383940   chr16:29383941-29484368         False   \n",
      "1   chr9:137103133-137103170  chr9:137103171-137103363         False   \n",
      "2    chr16:21404158-21404184   chr16:21404185-21404856         False   \n",
      "3     chr2:97200334-97200362    chr2:97200363-97209680         False   \n",
      "4    chr10:12155778-12155931   chr10:12155932-12156906         False   \n",
      "5   chr6:138926401-138926481  chr6:138921125-138926400         False   \n",
      "6   chr9:112701788-112701880  chr9:112694234-112701787         False   \n",
      "7    chr20:63701830-63701895   chr20:63700703-63701829         False   \n",
      "8   chr9:137103020-137103057  chr9:137103058-137103321         False   \n",
      "9    chr14:50174454-50174552   chr14:50161610-50174453         False   \n",
      "10  chr3:186506099-186506205                                    True   \n",
      "11  chr8:109213972-109260959                                    True   \n",
      "12  chr2:191745547-191830270                                    True   \n",
      "13  chr3:196662273-196669464                                    True   \n",
      "14  chr9:100395705-100436029                                    True   \n",
      "15  chr8:101715144-101734315                                    True   \n",
      "16  chr3:186502751-186502891                                   False   \n",
      "\n",
      "   exclusion_nmd  \n",
      "0                 \n",
      "1                 \n",
      "2                 \n",
      "3                 \n",
      "4                 \n",
      "5                 \n",
      "6                 \n",
      "7                 \n",
      "8                 \n",
      "9                 \n",
      "10                \n",
      "11                \n",
      "12                \n",
      "13                \n",
      "14                \n",
      "15                \n",
      "16                \n"
     ]
    }
   ],
   "source": [
    "database='/Users/rhythmicstar/projects/exon_evolution//gencode.v19.annotation.gtf.db'\n",
    "nmd_file='/Users/rhythmicstar/ipython_notebook/nmd/example_splicing.bed'\n",
    "event_ids_df = main(database, nmd_file)\n",
    "print(event_ids_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0\n",
      "\t 10\n",
      "\t 11\n",
      "\t 12\n",
      "1\n",
      "\t 10\n",
      "\t 11\n",
      "\t 12\n",
      "2\n",
      "\t 10\n",
      "\t 11\n",
      "\t 12\n",
      "3\n",
      "\t 10\n",
      "\t 11\n",
      "\t 12\n",
      "4\n",
      "\t 10\n",
      "\t 11\n",
      "\t 12\n"
     ]
    }
   ],
   "source": [
    "for i in range(5):\n",
    "    print(i)\n",
    "    for j in range(10, 15):\n",
    "        if j > 12:\n",
    "            break\n",
    "        print('\\t', j)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'0.8.6.1'"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gffutils.__version__"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[-100, -99, -98, -97]"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "[x-100 for x in range(4)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr14:50174454-50174552\n",
      "chr3:186506099-186506205\n",
      "Transcript\n",
      "True\n",
      "Transcript\n",
      "Transcript\n",
      "exon matches\n",
      "True\n",
      "Transcript\n",
      "True\n",
      "chr3:186506099-186506205\n",
      "chr3:186506099-186506205\n",
      "chr3:186506099-186506205\n",
      "chr3:186506099-186506205\n",
      "chr8:109213972-109260959\n",
      "chr8:109213972-109260959\n",
      "chr8:109213972-109260959\n",
      "chr8:109213972-109260959\n",
      "chr8:109213972-109260959\n",
      "chr8:109213972-109260959\n",
      "chr8:109213972-109260959\n",
      "chr8:109213972-109260959\n",
      "chr8:109213972-109260959\n",
      "chr8:109213972-109260959\n",
      "chr8:109213972-109260959\n",
      "chr8:109213972-109260959\n",
      "chr8:109213972-109260959\n",
      "chr8:109213972-109260959\n",
      "chr8:109213972-109260959\n",
      "chr8:109213972-109260959\n",
      "chr8:109213972-109260959\n",
      "chr8:109213972-109260959\n",
      "chr8:109213972-109260959\n",
      "chr8:109213972-109260959\n",
      "chr8:109213972-109260959\n",
      "chr8:109213972-109260959\n",
      "chr8:109213972-109260959\n",
      "chr8:109213972-109260959\n",
      "chr8:109213972-109260959\n",
      "chr8:109213972-109260959\n",
      "chr8:109213972-109260959\n",
      "chr8:109213972-109260959\n",
      "chr8:109213972-109260959\n",
      "chr8:109213972-109260959\n",
      "chr8:109213972-109260959\n",
      "chr8:109213972-109260959\n",
      "chr8:109213972-109260959\n",
      "chr8:109213972-109260959\n",
      "chr8:109213972-109260959\n",
      "chr8:109213972-109260959\n",
      "chr8:109213972-109260959\n",
      "chr8:109213972-109260959\n",
      "chr8:109213972-109260959\n",
      "chr8:109213972-109260959\n",
      "chr8:109213972-109260959\n",
      "chr8:109213972-109260959\n",
      "chr8:109213972-109260959\n",
      "chr8:109213972-109260959\n",
      "chr8:109213972-109260959\n",
      "chr8:109213972-109260959\n",
      "chr8:109213972-109260959\n",
      "chr8:109213972-109260959\n",
      "chr8:109213972-109260959\n",
      "chr8:109213972-109260959\n",
      "chr8:109213972-109260959\n",
      "chr8:109213972-109260959\n",
      "chr8:109213972-109260959\n",
      "chr8:109213972-109260959\n",
      "chr8:109213972-109260959\n",
      "chr8:109213972-109260959\n",
      "chr8:109213972-109260959\n",
      "chr8:109213972-109260959\n",
      "chr8:109213972-109260959\n",
      "chr8:109213972-109260959\n",
      "chr8:109213972-109260959\n",
      "chr8:109213972-109260959\n",
      "chr8:109213972-109260959\n",
      "chr8:109213972-109260959\n",
      "chr8:109213972-109260959\n",
      "chr8:109213972-109260959\n",
      "chr8:109213972-109260959\n",
      "chr8:109213972-109260959\n",
      "chr8:109213972-109260959\n",
      "chr8:109213972-109260959\n",
      "chr8:109213972-109260959\n",
      "chr8:109213972-109260959\n",
      "chr8:109213972-109260959\n",
      "chr8:109213972-109260959\n",
      "chr8:109213972-109260959\n",
      "chr8:109213972-109260959\n",
      "chr8:109213972-109260959\n",
      "chr8:109213972-109260959\n",
      "chr8:109213972-109260959\n",
      "chr8:109213972-109260959\n",
      "chr8:109213972-109260959\n",
      "chr8:109213972-109260959\n",
      "chr8:109213972-109260959\n",
      "chr8:109213972-109260959\n",
      "chr8:109213972-109260959\n",
      "chr8:109213972-109260959\n",
      "chr8:109213972-109260959\n",
      "chr8:109213972-109260959\n",
      "chr8:109213972-109260959\n",
      "chr8:109213972-109260959\n",
      "chr8:109213972-109260959\n",
      "chr8:109213972-109260959\n",
      "chr8:109213972-109260959\n",
      "chr8:109213972-109260959\n",
      "chr8:109213972-109260959\n",
      "chr8:109213972-109260959\n",
      "chr8:109213972-109260959\n",
      "chr8:109213972-109260959\n",
      "chr8:109213972-109260959\n",
      "chr8:109213972-109260959\n",
      "chr8:109213972-109260959\n",
      "chr8:109213972-109260959\n",
      "chr8:109213972-109260959\n",
      "chr8:109213972-109260959\n",
      "chr8:109213972-109260959\n",
      "chr8:109213972-109260959\n",
      "chr8:109213972-109260959\n",
      "chr8:109213972-109260959\n",
      "chr2:191745547-191830270\n",
      "chr2:191745547-191830270\n",
      "chr2:191745547-191830270\n",
      "chr2:191745547-191830270\n",
      "chr2:191745547-191830270\n",
      "chr2:191745547-191830270\n",
      "chr2:191745547-191830270\n",
      "chr2:191745547-191830270\n",
      "chr2:191745547-191830270\n",
      "chr2:191745547-191830270\n",
      "chr2:191745547-191830270\n",
      "chr2:191745547-191830270\n",
      "chr2:191745547-191830270\n",
      "chr2:191745547-191830270\n",
      "chr2:191745547-191830270\n",
      "chr2:191745547-191830270\n",
      "chr2:191745547-191830270\n",
      "chr2:191745547-191830270\n",
      "chr2:191745547-191830270\n",
      "chr2:191745547-191830270\n",
      "chr2:191745547-191830270\n",
      "chr2:191745547-191830270\n",
      "chr2:191745547-191830270\n",
      "chr2:191745547-191830270\n",
      "chr2:191745547-191830270\n",
      "chr2:191745547-191830270\n",
      "chr2:191745547-191830270\n",
      "chr2:191745547-191830270\n",
      "chr2:191745547-191830270\n",
      "chr2:191745547-191830270\n",
      "chr2:191745547-191830270\n",
      "chr2:191745547-191830270\n",
      "chr2:191745547-191830270\n",
      "chr2:191745547-191830270\n",
      "chr2:191745547-191830270\n",
      "chr2:191745547-191830270\n",
      "chr2:191745547-191830270\n",
      "chr2:191745547-191830270\n",
      "chr2:191745547-191830270\n",
      "chr2:191745547-191830270\n",
      "chr2:191745547-191830270\n",
      "chr2:191745547-191830270\n",
      "chr2:191745547-191830270\n",
      "chr2:191745547-191830270\n",
      "chr2:191745547-191830270\n",
      "chr2:191745547-191830270\n",
      "chr2:191745547-191830270\n",
      "chr2:191745547-191830270\n",
      "chr2:191745547-191830270\n",
      "chr2:191745547-191830270\n",
      "chr2:191745547-191830270\n",
      "chr2:191745547-191830270\n",
      "chr2:191745547-191830270\n",
      "chr2:191745547-191830270\n",
      "chr2:191745547-191830270\n",
      "chr2:191745547-191830270\n",
      "chr2:191745547-191830270\n",
      "chr2:191745547-191830270\n",
      "chr2:191745547-191830270\n",
      "chr2:191745547-191830270\n",
      "chr2:191745547-191830270\n",
      "chr2:191745547-191830270\n",
      "chr2:191745547-191830270\n",
      "chr2:191745547-191830270\n",
      "chr2:191745547-191830270\n",
      "chr2:191745547-191830270\n",
      "chr2:191745547-191830270\n",
      "chr2:191745547-191830270\n",
      "chr2:191745547-191830270\n",
      "chr2:191745547-191830270\n",
      "chr2:191745547-191830270\n",
      "chr2:191745547-191830270\n",
      "chr2:191745547-191830270\n",
      "chr2:191745547-191830270\n",
      "chr2:191745547-191830270\n",
      "chr2:191745547-191830270\n",
      "chr2:191745547-191830270\n",
      "chr2:191745547-191830270\n",
      "chr2:191745547-191830270\n",
      "chr2:191745547-191830270\n",
      "chr2:191745547-191830270\n",
      "chr2:191745547-191830270\n",
      "chr2:191745547-191830270\n",
      "chr2:191745547-191830270\n",
      "chr2:191745547-191830270\n",
      "chr2:191745547-191830270\n",
      "chr2:191745547-191830270\n",
      "chr2:191745547-191830270\n",
      "chr2:191745547-191830270\n",
      "chr2:191745547-191830270\n",
      "chr2:191745547-191830270\n",
      "chr2:191745547-191830270\n",
      "chr2:191745547-191830270\n",
      "chr2:191745547-191830270\n",
      "chr2:191745547-191830270\n",
      "chr2:191745547-191830270\n",
      "chr2:191745547-191830270\n",
      "chr2:191745547-191830270\n",
      "chr3:196662273-196669464\n",
      "chr3:196662273-196669464\n",
      "chr3:196662273-196669464\n",
      "chr3:196662273-196669464\n",
      "chr3:196662273-196669464\n",
      "chr3:196662273-196669464\n",
      "chr3:196662273-196669464\n",
      "chr3:196662273-196669464\n",
      "chr3:196662273-196669464\n",
      "chr3:196662273-196669464\n",
      "chr3:196662273-196669464\n",
      "chr3:196662273-196669464\n",
      "chr3:196662273-196669464\n",
      "chr3:196662273-196669464\n",
      "chr3:196662273-196669464\n",
      "chr3:196662273-196669464\n",
      "chr3:196662273-196669464\n",
      "chr3:196662273-196669464\n",
      "chr3:196662273-196669464\n",
      "chr3:196662273-196669464\n",
      "chr3:196662273-196669464\n",
      "chr3:196662273-196669464\n",
      "chr3:196662273-196669464\n",
      "chr3:196662273-196669464\n",
      "chr3:196662273-196669464\n",
      "chr3:196662273-196669464\n",
      "chr3:196662273-196669464\n",
      "chr3:196662273-196669464\n",
      "chr3:196662273-196669464\n",
      "chr3:196662273-196669464\n",
      "chr3:196662273-196669464\n",
      "chr3:196662273-196669464\n",
      "chr3:196662273-196669464\n",
      "chr3:196662273-196669464\n",
      "chr3:196662273-196669464\n",
      "chr3:196662273-196669464\n",
      "chr3:196662273-196669464\n",
      "chr3:196662273-196669464\n",
      "chr3:196662273-196669464\n",
      "chr3:196662273-196669464\n",
      "chr3:196662273-196669464\n",
      "chr3:196662273-196669464\n",
      "chr3:196662273-196669464\n",
      "chr3:196662273-196669464\n",
      "chr3:196662273-196669464\n",
      "chr3:196662273-196669464\n",
      "chr9:100395705-100436029\n",
      "chr9:100395705-100436029\n",
      "chr9:100395705-100436029\n",
      "chr9:100395705-100436029\n",
      "chr9:100395705-100436029\n",
      "chr9:100395705-100436029\n",
      "chr9:100395705-100436029\n",
      "chr9:100395705-100436029\n",
      "chr9:100395705-100436029\n",
      "chr9:100395705-100436029\n",
      "chr9:100395705-100436029\n",
      "chr9:100395705-100436029\n",
      "chr9:100395705-100436029\n",
      "chr9:100395705-100436029\n",
      "chr9:100395705-100436029\n",
      "chr9:100395705-100436029\n",
      "chr9:100395705-100436029\n",
      "chr9:100395705-100436029\n",
      "chr9:100395705-100436029\n",
      "chr9:100395705-100436029\n",
      "chr9:100395705-100436029\n",
      "chr9:100395705-100436029\n",
      "chr9:100395705-100436029\n",
      "chr9:100395705-100436029\n",
      "chr9:100395705-100436029\n",
      "chr9:100395705-100436029\n",
      "chr9:100395705-100436029\n",
      "chr9:100395705-100436029\n",
      "chr9:100395705-100436029\n",
      "chr9:100395705-100436029\n",
      "chr9:100395705-100436029\n",
      "chr9:100395705-100436029\n",
      "chr9:100395705-100436029\n",
      "chr9:100395705-100436029\n",
      "chr9:100395705-100436029\n",
      "chr9:100395705-100436029\n",
      "chr9:100395705-100436029\n",
      "chr9:100395705-100436029\n",
      "chr9:100395705-100436029\n",
      "chr9:100395705-100436029\n",
      "chr8:101715144-101734315\n",
      "chr8:101715144-101734315\n",
      "chr8:101715144-101734315\n",
      "chr8:101715144-101734315\n",
      "chr8:101715144-101734315\n",
      "chr8:101715144-101734315\n",
      "chr8:101715144-101734315\n",
      "chr8:101715144-101734315\n",
      "chr8:101715144-101734315\n",
      "chr8:101715144-101734315\n",
      "chr8:101715144-101734315\n",
      "chr8:101715144-101734315\n",
      "chr8:101715144-101734315\n",
      "chr8:101715144-101734315\n",
      "chr8:101715144-101734315\n",
      "chr8:101715144-101734315\n",
      "chr8:101715144-101734315\n",
      "chr8:101715144-101734315\n",
      "chr8:101715144-101734315\n",
      "chr8:101715144-101734315\n",
      "chr8:101715144-101734315\n",
      "chr8:101715144-101734315\n",
      "chr8:101715144-101734315\n",
      "chr8:101715144-101734315\n",
      "chr8:101715144-101734315\n",
      "chr8:101715144-101734315\n",
      "chr8:101715144-101734315\n",
      "chr8:101715144-101734315\n",
      "chr8:101715144-101734315\n",
      "chr8:101715144-101734315\n",
      "chr8:101715144-101734315\n",
      "chr8:101715144-101734315\n",
      "chr8:101715144-101734315\n",
      "chr8:101715144-101734315\n",
      "chr8:101715144-101734315\n",
      "chr8:101715144-101734315\n",
      "chr8:101715144-101734315\n",
      "chr8:101715144-101734315\n",
      "chr8:101715144-101734315\n",
      "chr8:101715144-101734315\n",
      "chr8:101715144-101734315\n",
      "chr8:101715144-101734315\n",
      "chr8:101715144-101734315\n",
      "chr8:101715144-101734315\n",
      "chr8:101715144-101734315\n",
      "chr8:101715144-101734315\n",
      "chr8:101715144-101734315\n",
      "chr8:101715144-101734315\n",
      "chr8:101715144-101734315\n",
      "chr8:101715144-101734315\n",
      "chr8:101715144-101734315\n",
      "chr8:101715144-101734315\n",
      "chr8:101715144-101734315\n",
      "chr8:101715144-101734315\n",
      "chr8:101715144-101734315\n",
      "chr8:101715144-101734315\n",
      "chr8:101715144-101734315\n",
      "chr8:101715144-101734315\n",
      "chr8:101715144-101734315\n",
      "chr8:101715144-101734315\n",
      "chr8:101715144-101734315\n",
      "chr8:101715144-101734315\n",
      "chr8:101715144-101734315\n",
      "chr8:101715144-101734315\n",
      "chr8:101715144-101734315\n",
      "chr8:101715144-101734315\n",
      "chr8:101715144-101734315\n",
      "chr8:101715144-101734315\n",
      "chr8:101715144-101734315\n",
      "chr8:101715144-101734315\n",
      "chr8:101715144-101734315\n",
      "chr8:101715144-101734315\n",
      "chr8:101715144-101734315\n",
      "chr8:101715144-101734315\n",
      "chr8:101715144-101734315\n",
      "chr8:101715144-101734315\n",
      "chr8:101715144-101734315\n",
      "chr8:101715144-101734315\n",
      "chr8:101715144-101734315\n",
      "chr8:101715144-101734315\n",
      "chr8:101715144-101734315\n",
      "chr8:101715144-101734315\n",
      "chr8:101715144-101734315\n",
      "chr8:101715144-101734315\n",
      "chr8:101715144-101734315\n",
      "chr8:101715144-101734315\n",
      "chr8:101715144-101734315\n",
      "chr8:101715144-101734315\n",
      "chr8:101715144-101734315\n",
      "chr8:101715144-101734315\n",
      "chr8:101715144-101734315\n",
      "chr8:101715144-101734315\n",
      "chr8:101715144-101734315\n",
      "chr8:101715144-101734315\n",
      "chr8:101715144-101734315\n",
      "chr8:101715144-101734315\n",
      "chr8:101715144-101734315\n",
      "chr8:101715144-101734315\n",
      "chr8:101715144-101734315\n",
      "chr8:101715144-101734315\n",
      "chr8:101715144-101734315\n",
      "chr8:101715144-101734315\n",
      "chr8:101715144-101734315\n",
      "chr8:101715144-101734315\n",
      "chr8:101715144-101734315\n",
      "chr8:101715144-101734315\n",
      "chr8:101715144-101734315\n",
      "chr8:101715144-101734315\n",
      "chr8:101715144-101734315\n",
      "chr8:101715144-101734315\n",
      "chr8:101715144-101734315\n",
      "chr8:101715144-101734315\n",
      "chr8:101715144-101734315\n",
      "chr8:101715144-101734315\n",
      "chr8:101715144-101734315\n",
      "chr8:101715144-101734315\n",
      "chr8:101715144-101734315\n",
      "chr8:101715144-101734315\n",
      "chr8:101715144-101734315\n",
      "chr8:101715144-101734315\n",
      "chr8:101715144-101734315\n",
      "chr8:101715144-101734315\n",
      "chr8:101715144-101734315\n",
      "chr8:101715144-101734315\n",
      "chr8:101715144-101734315\n",
      "chr8:101715144-101734315\n",
      "chr8:101715144-101734315\n",
      "chr8:101715144-101734315\n",
      "chr8:101715144-101734315\n",
      "chr8:101715144-101734315\n",
      "chr3:186502751-186502891\n",
      "chr3:186502751-186502891\n",
      "chr3:186502751-186502891\n",
      "chr3:186502751-186502891\n",
      "chr3:186502751-186502891\n",
      "chr3:186502751-186502891\n",
      "chr3:186502751-186502891\n",
      "chr3:186502751-186502891\n",
      "chr3:186502751-186502891\n",
      "chr3:186502751-186502891\n",
      "chr3:186502751-186502891\n",
      "                    isoform1                  isoform2  \\\n",
      "0    chr16:29383506-29484368   chr16:29383506-29383883   \n",
      "1   chr9:137102499-137103363  chr9:137102499-137103132   \n",
      "2    chr16:21404089-21404856   chr16:21404089-21404157   \n",
      "3     chr2:97198492-97209680    chr2:97198492-97200333   \n",
      "4    chr10:12149962-12156906   chr10:12149962-12155777   \n",
      "5   chr6:138921125-138929976  chr6:138926482-138929976   \n",
      "6   chr9:112694234-112716460  chr9:112701881-112716460   \n",
      "7    chr20:63700703-63702135   chr20:63701896-63702135   \n",
      "8   chr9:137102898-137103321  chr9:137102898-137103019   \n",
      "9    chr14:50161610-50180571   chr14:50174553-50180571   \n",
      "10                                                       \n",
      "11                                                       \n",
      "12                                                       \n",
      "13                                                       \n",
      "14                                                       \n",
      "15                                                       \n",
      "16                                                       \n",
      "\n",
      "                        exon                  junction inclusion_nmd  \\\n",
      "0    chr16:29383884-29383940   chr16:29383941-29484368                 \n",
      "1   chr9:137103133-137103170  chr9:137103171-137103363                 \n",
      "2    chr16:21404158-21404184   chr16:21404185-21404856                 \n",
      "3     chr2:97200334-97200362    chr2:97200363-97209680                 \n",
      "4    chr10:12155778-12155931   chr10:12155932-12156906                 \n",
      "5   chr6:138926401-138926481  chr6:138921125-138926400                 \n",
      "6   chr9:112701788-112701880  chr9:112694234-112701787                 \n",
      "7    chr20:63701830-63701895   chr20:63700703-63701829                 \n",
      "8   chr9:137103020-137103057  chr9:137103058-137103321                 \n",
      "9    chr14:50174454-50174552   chr14:50161610-50174453                 \n",
      "10  chr3:186506099-186506205                                           \n",
      "11  chr8:109213972-109260959                                           \n",
      "12  chr2:191745547-191830270                                           \n",
      "13  chr3:196662273-196669464                                           \n",
      "14  chr9:100395705-100436029                                           \n",
      "15  chr8:101715144-101734315                                           \n",
      "16  chr3:186502751-186502891                                           \n",
      "\n",
      "   exclusion_nmd  \n",
      "0                 \n",
      "1                 \n",
      "2                 \n",
      "3                 \n",
      "4                 \n",
      "5                 \n",
      "6                 \n",
      "7                 \n",
      "8                 \n",
      "9                 \n",
      "10          True  \n",
      "11                \n",
      "12                \n",
      "13                \n",
      "14                \n",
      "15                \n",
      "16                \n"
     ]
    }
   ],
   "source": [
    "database='/Users/rhythmicstar/projects/exon_evolution//gencode.v19.annotation.gtf.db'\n",
    "nmd_file='/Users/rhythmicstar/ipython_notebook/nmd/example_splicing.bed'\n",
    "event_ids_df = possible_nmd(database, nmd_file)\n",
    "DBHuman = gffutils.FeatureDB(database, keep_order=True)\n",
    "for i, row in event_ids_df.iterrows():\n",
    "    nmd = False\n",
    "    location = row['exon']\n",
    "    for index, exon in enumerate(DBHuman.features_of_type(('exon'), limit=location)):\n",
    "        print(location)\n",
    "        if (location==(\"{chrom}:{start}-{end}\".format(chrom=exon.chrom, start=exon.start, end=exon.end))):\n",
    "            for transcript in DBHuman.children(exon['gene_id'][0], featuretype='transcript'): #find transcripts that exon is not in\n",
    "                exonInTrans = False\n",
    "                if (transcript['transcript_type']==[\"nonsense_mediated_decay\"]):\n",
    "                    print(\"Transcript\")\n",
    "                    for human_exon in DBHuman.children(transcript, featuretype='exon'): #look in all exons after the missing exon\n",
    "                        if (\"{chrom}:{start}-{end}\".format(chrom=exon.chrom, start=exon.start, end=exon.end)==\"{chrom}:{start}-{end}\"\n",
    "                            .format(chrom=human_exon.chrom, start=human_exon.start, end=human_exon.end)): #if transcript contains the exon\n",
    "                            print(\"exon matches\")\n",
    "                            break #is this correct? Want it to go to the next transcript\n",
    "\n",
    "                    for human_exon in DBHuman.children(transcript, featuretype='exon'): #look in all exons after the missing exon\n",
    "                        for stop_codon in DBHuman.children(human_exon['gene_id'][0], featuretype='stop_codon'):\n",
    "                            if (human_exon.strand == '+' and (human_exon.start >= int(event_ids_df.ix[i, 'exon'].split(':')[1].split('-')[1]))): #if strand is +, all exons with bigger location\n",
    "                                if (stop_codon_nmd(human_exon, stop_codon)==True):\n",
    "                                    print(\"True\")\n",
    "                                    event_ids_df.ix[i, 5] = 'True'\n",
    "                                    break\n",
    "                            elif (exon.strand == '-' and (exon.end <= int(event_ids_df.ix[row, 'exon'].split(':')[1].split('-')[1]))): #if strand is -, all exons w/ small location\n",
    "                                if (stop_codon_nmd(human_exon, stop_codon)==True):\n",
    "                                    print(\"True\")\n",
    "                                    event_ids_df.ix[i, 5] = 'True'\n",
    "                                    break\n",
    "print(event_ids_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "event_ids_df.ix[i, 5] = 'False'"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
